test/mas_test.cpp

```diff
@@ -44,54 +46,45 @@ struct edge_t
 };
 
 template < typename Graph, typename KeyedUpdatablePriorityQueue >
-class mas_edge_connectivity_visitor : public boost::default_mas_visitor
+class mas_test_visitor : public boost::default_mas_visitor
 {
 public:
     typedef typename boost::graph_traits< Graph >::vertex_descriptor
         vertex_descriptor;
     typedef typename KeyedUpdatablePriorityQueue::key_type weight_type;
-#if 0
-    mas_edge_connectivity_visitor(const mas_edge_connectivity_visitor<Graph, KeyedUpdatablePriorityQueue>& r)
-      : m_pq(r.m_pq), m_curr(r.m_curr), m_prev(r.m_prev),
-        m_reach_weight(r.m_reach_weight) {
-          BOOST_TEST_MESSAGE( "COPY CTOR" );
-        }
-#endif
-    explicit mas_edge_connectivity_visitor(KeyedUpdatablePriorityQueue& pq)
-    : m_pq(pq)
-    , m_curr(new vertex_descriptor(0))
-    , m_prev(new vertex_descriptor(0))
-    , m_reach_weight(new weight_type(0))
-    {
-        //    BOOST_TEST_MESSAGE( "CTOR" );
-    }
+
+    explicit mas_test_visitor(KeyedUpdatablePriorityQueue& pq)
+    :   m_pq_(pq),
+        vertex_visit_order_(boost::make_shared< std::vector< vertex_descriptor > >()),
+        vertex_weights_when_visited_(boost::make_shared< std::vector< weight_type > >())
+    {}
 
     void clear()
     {
-        *m_curr = 0;
-        *m_prev = 0;
-        *m_reach_weight = 0;
+        vertex_visit_order_->clear();
+        vertex_weights_when_visited_->clear();
     }
 
-    // template <typename Vertex> //, typename Graph>
-    // void start_vertex(Vertex u, const Graph& g) {
     void start_vertex(vertex_descriptor u, const Graph& g)
     {
-        *m_prev = *m_curr;
-        *m_curr = u;
-        // BOOST_TEST_MESSAGE( "Initializing Vertex(weight): " << u << "(" <<
-        // *m_reach_weight << ")" );
-        *m_reach_weight = get(m_pq.keys(), u);
+        vertex_visit_order_->push_back(u);
+
+        const weight_type u_weight = get(m_pq_.keys(), u);
+        vertex_weights_when_visited_->push_back(u_weight);
     }
 
-    vertex_descriptor curr() const { return *m_curr; }
-    vertex_descriptor prev() const { return *m_prev; }
-    weight_type reach_weight() const { return *m_reach_weight; }
+    const std::vector<vertex_descriptor>& vertex_visit_order() const {
+        return *vertex_visit_order_;
+    }
+
+    const std::vector<weight_type>& vertex_weights_when_visited() const {
+        return *vertex_weights_when_visited_;
+    }
 
 private:
-    const KeyedUpdatablePriorityQueue& m_pq;
-    boost::shared_ptr< vertex_descriptor > m_curr, m_prev;
-    boost::shared_ptr< weight_type > m_reach_weight;
+    const KeyedUpdatablePriorityQueue& m_pq_;
+    boost::shared_ptr< std::vector< vertex_descriptor > > vertex_visit_order_;
+    boost::shared_ptr< std::vector< weight_type > > vertex_weights_when_visited_;
```

Comment:
Do these have to be `shared_ptr`? I don't have time to look at the full code, but I couldn't see a reason for it here in the PR code.

Reply:
These are `shared_ptr`'s because the visitor object is passed by value during the call to `maximum_adjacency_search`. Apparently this passing by value is 'normal' as I'm also seeing it in BFS.

In DFS however, it's passing by reference, which makes more sense to me. This might be a separate future work itme.

Reply:
Ah, I see. Yeah, I think those algorithms should probably take the visitor by value _and_ return it by value.